llvm/lib/SYCLLowerIR/LowerSPIRVAddressSpaceCasts.cpp

```diff
@@ -0,0 +1,93 @@
+#include "llvm/SYCLLowerIR/LowerSPIRVAddressSpaceCasts.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+
+namespace llvm {
+
+static bool visitCallInst(CallInst *CI) {
+  Function *Callee = CI->getCalledFunction();
+  if (!Callee || !Callee->isDeclaration())
+    return false;
+
+  // We are looking for this pattern where an LLVM addrspace cast
+  // is followed by a call to a SPIR-V intrinsic:
+  // %0 = addrspacecast ptr addrspace(1) %src to ptr addrspace(4)
+  // %1 = call ptr addrspace(1) @cast(ptr addrspace(4) %0, ...)
+  if (Callee->getName() != "_Z41__spirv_GenericCastToPtrExplicit_ToGlobalPvi" &&
+      Callee->getName() != "_Z40__spirv_GenericCastToPtrExplicit_ToLocalPvi" &&
+      Callee->getName() != "_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi")
+    return false;
+
+  // Check if the first argument is an addrspacecast instruction
+  Value *Arg = CI->getArgOperand(0);
+  auto *DstType = cast<PointerType>(CI->getType());
+  if (auto *ASC = dyn_cast<AddrSpaceCastInst>(Arg)) {
+    Value *ReplacedValue = nullptr;
+    if (ASC->getSrcAddressSpace() == DstType->getAddressSpace()) {
+      ReplacedValue = ASC->getOperand(0);
+
+      // Check if the addrspacecast argument is a null pointer constant,
+      // and if so, we can replace the call with a null pointer constant.
+      if (isa<ConstantPointerNull>(ReplacedValue))
+        ReplacedValue = ConstantPointerNull::get(PointerType::get(
+            DstType->getContext(), DstType->getAddressSpace()));
+
+    } else {
+      assert(ASC->getSrcAddressSpace() != 4);
```

Comment:
+1 to @steffenlarsen's comment about outlining this magic number. Even if it is not made into a parameter, it should at least be made a named constant within the file so it is less magic